rust/lance/src/dataset/write/merge_insert/exec/write.rs

```diff
@@ -167,8 +179,118 @@ impl FullSchemaMergeInsertExec {
         input_stream: SendableRecordBatchStream,
         merge_state: Arc<Mutex<MergeState>>,
     ) -> DFResult<SendableRecordBatchStream> {
-        let input_schema = input_stream.schema();
+        let enable_stable_row_ids = {
+            let state = merge_state.lock().map_err(|e| {
+                datafusion::error::DataFusionError::Internal(format!(
+                    "Failed to lock merge state: {}",
+                    e
+                ))
+            })?;
+            state.enable_stable_row_ids
+        };
+
+        if enable_stable_row_ids {
+            self.create_ordered_update_insert_stream(input_stream, merge_state)
+        } else {
+            self.create_streaming_write_stream(input_stream, merge_state)
+        }
+    }
+
+    /// High-performance streaming implementation for non-stable row ID scenarios
+    ///
+    /// It processes batches one at a time as they arrive from the input stream,
+    /// immediately filtering and transforming each batch without buffering.
+    fn create_streaming_write_stream(
+        &self,
+        input_stream: SendableRecordBatchStream,
+        merge_state: Arc<Mutex<MergeState>>,
+    ) -> DFResult<SendableRecordBatchStream> {
+        let (_, rowaddr_idx, action_idx, data_column_indices, output_schema) =
+            self.prepare_stream_schema(input_stream.schema())?;
 
+        let output_schema_clone = output_schema.clone();
+        let stream = input_stream.map(move |batch_result| -> DFResult<RecordBatch> {
+            let batch = batch_result?;
+            let (row_addr_array, action_array) =
+                Self::extract_control_arrays(&batch, rowaddr_idx, action_idx)?;
+
+            // Process each row using the shared state
+            let mut keep_rows: Vec<u32> = Vec::new();
```

Comment:
For the common case of upsert, I would pre-size this vec:


```suggestion
            let mut keep_rows: Vec<u32> = Vec::with_capacity(batch.num_rows());
```